crates/sui-core/src/consensus_throughput_calculator.rs

```diff
@@ -0,0 +1,623 @@
+        let observations_window = observations_window
+            .unwrap_or(NonZeroU64::new(DEFAULT_OBSERVATIONS_WINDOW).unwrap())
+            .get();
+
+        Self {
+            observations_window,
+            inner: Mutex::new(ConsensusThroughputCalculatorInner::default()),
+            current_throughput: ArcSwap::from_pointee((0, 0)),
+            metrics,
+        }
+    }
+
+    // Adds an observation of the number of transactions that have been sequenced after deduplication
+    // and the corresponding leader timestamp. The observation timestamps should be monotonically
+    // incremented otherwise observation will be ignored.
+    pub fn add_transactions(&self, timestamp_ms: TimestampMs, num_of_transactions: u64) {
+        let mut inner = self.inner.lock();
+        let timestamp_secs: TimestampSecs = timestamp_ms / 1_000; // lowest bucket we care is seconds
+
+        if let Some((front_ts, transactions)) = inner.observations.pop_front() {
+            // First check that the timestamp is monotonically incremented - ignore any observation that is not
+            // later from previous one (it shouldn't really happen).
+            if timestamp_secs < front_ts {
+                warn!("Ignoring observation of transactions:{} as has earlier timestamp than last observation {}s < {}s", num_of_transactions, timestamp_secs, front_ts);
+                return;
+            }
+
+            // Not very likely, but if transactions refer to same second we add to the last element.
+            if timestamp_secs == front_ts {
+                inner
+                    .observations
+                    .push_front((front_ts, transactions + num_of_transactions));
+            } else {
+                inner.observations.push_front((front_ts, transactions));
+                inner
+                    .observations
+                    .push_front((timestamp_secs, num_of_transactions));
+            }
+        } else {
+            inner
+                .observations
+                .push_front((timestamp_secs, num_of_transactions));
+        }
+
+        // update total number of transactions in the observations list
+        inner.total_transactions = inner.total_transactions.saturating_add(num_of_transactions);
+
+        // If we have more values on our window of max values, remove the last one, and update the num of transactions
+        // We also update the throughput profile when we have at least observations_window values in our observations.
+        if inner.observations.len() as u64 > self.observations_window {
+            let (last_element_ts, last_element_transactions) =
+                inner.observations.pop_back().unwrap();
+            inner.total_transactions = inner
+                .total_transactions
+                .saturating_sub(last_element_transactions);
+
+            // get the first element's timestamp to calculate the transaction rate
```

Comment:
Why not always update throughput when there are enough observations? We know the total transactions have been updated.

Reply:
Hmm not sure what you mean but that's what we do here. Once we have enough observations - `observations_window` elements - we'll be continuously calculating the throughput.

Reply:
This branch is guarded by `if inner.observations.len() as u64 > self.observations_window`, so when `inner.observations.len() == self.observations_window`, the measured throughput does not get updated. This happens when an observation updates an existing bucket. It seems unintuitive that only observation that creates a new bucket updates the total throughput.

Reply:
Ah ok I see the point. Yeah so this is by design - I mean that I thought about this case - and chose the above behaviour. Let's consider the approach where I do not update the last bucket when timestamps equivocate . Now let's imagine that I am adding observations only with the same timestamp. So buckets would accumulate and eventually they fill up `observation_window` elements. Now, when I'll try to calculate throughput, since the `last` and the `first` element will have the _exact_ same timestamp, the time period diff will be `0` - hence I won't be able to calculate the throughput. I know that those are edge cases, but when thinking about it I couldn't find a solution that would satisfy all the cases. So my point is that observations that fall _under the exact same timestamp_ (and we are talking about miliseconds here) it's as if they should be reported all together in first place.

Now practically I do not really expect to have many occurrences where we update the existing bucket - that would mean that consensus for monotonically increasing commits is not producing monotonically increasing timestamps.

Reply:
When re-measuring throughput, you don't need to add a new "bucket" when new observation timestamp is the same as the last. You can keep the existing logic to modify the total # of txns in the last bucket. And throughput can be calculated when there are enough buckets, so there shouldn't be a division by 0 problem.

Having two commits within 1s should be much more common than actual equivocations. Current commit interval is ~1s. And during equivocations worse problems can happen presumably.

But this is just something I noticed that seems a bit unusual. It probably will not matter much in practice. If you insist on the current logic, it is fine.